src/wasm/wat-parser.cpp

```diff
@@ -1610,10 +1627,65 @@ MaybeResult<typename Ctx::InstrT> instr(Ctx& ctx, ParseInput& in) {
 template<typename Ctx>
 Result<typename Ctx::InstrsT> instrs(Ctx& ctx, ParseInput& in) {
   auto insts = ctx.makeInstrs();
-  while (auto inst = instr(ctx, in)) {
-    CHECK_ERR(inst);
-    ctx.appendInstr(insts, *inst);
+
+  while (true) {
+    if (in.takeLParen()) {
+      // A stack of (start, end) position pairs defining the positions of
+      // instructions that need to be parsed after their folded children.
+      std::vector<std::pair<Index, std::optional<Index>>> foldedInstrs;
+
+      // Begin a folded instruction. Push its start position and a placeholder
+      // end position.
+      foldedInstrs.push_back({in.getPos(), {}});
+      while (!foldedInstrs.empty()) {
+        // Consume everything up to the next paren. This span will be parsed as
+        // an instruction later after its folded children have been parsed.
+        if (!in.takeUntilParen()) {
+          return ParseInput(in, foldedInstrs.back().first)
+            .err("unterminated folded instruction");
+        }
+
+        if (!foldedInstrs.back().second) {
+          // The folded instruction we just started should end here.
+          foldedInstrs.back().second = in.getPos();
+        }
+
+        // We have either the start of a new folded child or the end of the last
+        // one.
+        if (in.takeLParen()) {
+          foldedInstrs.push_back({in.getPos(), {}});
+        } else if (in.takeRParen()) {
+          auto [start, end] = foldedInstrs.back();
+          assert(end && "Should have found end of instruction");
+          foldedInstrs.pop_back();
+
+          ParseInput foldedIn(in, start);
+          if (auto inst = instr(ctx, foldedIn)) {
```

Comment:
I'm probably missing something, but I'd expect to see a code path that handles non-folded children in a folded instruction around here, which should call `instrs()` on a substring of the folded instruction. Is that right? If so where is it?

What I mean is this:
```
(block ;; after parsing the shape of the block, where is the code that will parse the nops?
  nop
  nop
  nop
)
```

Reply:
Blocks and other control flow structures will definitely need to be special cased in this code once we support them at all. The two locations here that check for left parens will have to be updated to check for control flow structures first. Since we don't support control flow structures yet, it's not possible to have non-folded instructions as children of folded instructions. This surprised me when I was reading the spec, but it definitely makes the parsing problem simpler.